framework/src/variables/MooseVariableField.C

```diff
@@ -85,7 +85,11 @@ template <typename OutputType>
 bool
 MooseVariableField<OutputType>::isArray() const
 {
-  return std::is_same<OutputType, RealEigenVector>::value;
+  const auto is_array = MooseVariableBase::isArray();
+  if (std::is_same<OutputType, RealEigenVector>::value != is_array)
```

Comment:
```suggestion
  if constexpr (std::is_same<OutputType, RealEigenVector>::value != is_array)
```

WIll make this a compile check

Reply:
I get a compiler error when trying this suggested code change because 
```
 note: initializer of 'is_array' is not a constant expression
 ```
 I don't think this can work as suggested, but I can do something similar. This part, `std::is_same<OutputType, RealEigenVector>::value`, at least can have `constexpr`.

Reply:
Oh wait, hold on... why do you need a runtime check now? Is this related to why you have to add the virtual method?

Reply:
Okay, that's fine. I follow.